tests/test_libs/test_bucket.py

```diff
@@ -0,0 +1,216 @@
+        )
+
+    def fake_grants(self):
+        obj = mock.Mock()
+        obj.Acl.return_value.grants = [
+            self.fake_grant_name("Group"),
+            self.fake_grant_name("CanonicalUser"),
+        ]
+        return obj
+
+    def test_grant(self):
+        assert bucket.get_grants(self.fake_grants()) == [
+            ["red-deer", "FULL CONTROL"],
+            ["black-deer", "FULL CONTROL"],
+        ]
+
+    def test_gmt_policy(self, monkeypatch):
+        monkeypatch.setattr(gmt, "policy_id", lambda name, auth: None)
+        monkeypatch.setattr(
+            gmt, "policy_description", lambda id: "Replica Data to all nodes"
+        )
+        assert (
+            bucket.bucket_gmt_policy("bucket-name", "auth")
+            == "Replica Data to all nodes"
+        )
+
+    def fake_client(self):
+        client = mock.Mock()
+        client.Bucket.return_value.meta.client = None
+        return client
+
+    def test_bucket_info(self, monkeypatch):
+        monkeypatch.setattr(
+            bucket, "bucket_gmt_policy", lambda name, auth: "Replica Data to all nodes"
+        )
+        monkeypatch.setattr(bucket, "get_grants", lambda bucket: "123")
+        monkeypatch.setattr(bucket, "get_cors", lambda bucket: "None")
+        monkeypatch.setattr(bucket, "get_policy", lambda bucket: "None")
+        monkeypatch.setattr(bucket, "get_expiration", lambda client, bucket: "None")
+        monkeypatch.setattr(bucket, "get_location", lambda client, bucket: "Jakarta")
+
+        assert bucket.bucket_info(self.fake_client(), "bucket-name", "auth") == {
+            "ACL": "123",
+            "CORS": "None",
+            "Policy": "None",
+            "Expiration": "None",
+            "Location": "Jakarta",
+            "GmtPolicy": "Replica Data to all nodes",
+        }
+
+    def fake_object_resource(self):
+        obj = mock.Mock()
+        obj.Object.return_value.storage_class = "Group"
+        obj.Object.return_value.content_type = "image/png"
+        obj.Object.return_value.content_length = 1280213123
+        obj.Object.return_value.last_modified = "2019-09-24 01:01:00"
+        obj.Object.return_value.e_tag = "967bd45c88a91f829755742752fe9932"
+        return obj
+
+    def test_object_info(self, monkeypatch):
+        monkeypatch.setattr(
+            bucket, "get_grants", lambda obj: ["red-deer", "FULL CONTROL"]
+        )
+        assert bucket.object_info(
+            self.fake_object_resource(), "bucket-name", "obj-name"
+        ) == {
+            "ACL": ["red-deer", "FULL CONTROL"],
+            "Size": 1280213123,
+            "LastModified": "2019-09-24 01:01:00",
+            "MD5": "967bd45c88a91f829755742752fe9932",
+            "MimeType": "image/png",
+            "StorageClass": "Group",
+        }
+
+    def fake_url(self):
+        client = mock.Mock()
+        client.meta.client.generate_presigned_url.return_value = (
```

Comment:
it's safer to use random url. not ours. expose as little as possible our internal data.